### DISCUSSION TITLE
Storer to mimic XA transactions

### DISCUSSION THREAD

**Discussion by guykatz** [microstream-one/microstream#537] [Q&A] [answered]
Hi all;
in my aplication i have several modules storing data under an XA transaction (2 phase commit)
using microstream I need to:
- mark the start of storing process
- store many things from different place :)
- either store or rollback
I assume I can achieve this using storer commit and clean? (is clean the rollback to the commit? or its something else?
any suggestions on how to achieve what I need with microstream?

**Comment by hg-ms** [comment, marked as answer]
Hi,
From our docu regarding [transaction](https://docs.microstream.one/manual/storage/storing-data/transactions.html#_transactions):
> MicroStream does not provide explicit transactions, every call to a store method is automatically a transaction.
> A store operation is an atomic all or nothing operation
> If the store call is successful all data is written to the storage. Otherwise, no data is persisted. Partially persisted data will be reverted.

To mimic a XA transaction you can use an explicit Storer. 
` Storer storer = storageManager.createStorer()`
 That way you have two phases:
 
1.	Data Collection
After creating a new storer it has no data to store. To “collect” that data use the store() and storeAll() methods of the storer. You can do more than one ’store()’.
If you decide to abort before persisting data, you can either call the ‘clear()’ method or just let the Java-GC delete the Storer.
The clear() method will reset the Storer to an empty state, so you can reuse it..

2.	Persisting
When collecting data is done, call commit() to persist everything at once. This call ether succeeds or fails with an exception. In case of partially written data the storage system will revert that. A successful commit will also A successful commit will clean the storer too.
